### PR TITLE
Bump the minimum Java version to Java 11

### PR DESCRIPTION
**Pull request by jasontedor**
With the 8.0.0 release of Elasticsearch we will bump the minimum required Java to Java 11. This commit puts this into effect on the master branch.

### CODE REVIEW

**Comment by jasontedor** [comment]
@jkakavas FYI this means that we will need a `java11fips` JVM in CI to continue supporting FIPS.

**Comment by jasontedor** [comment]
@danielmitterdorfer This will disrupt performance benchmarks, how do you want to proceed?

File: docs/reference/migration/migrate_8_0/packaging.asciidoc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[float]
+[[breaking_80_packaging_changes]]
+=== Packaging changes
+
+[float]
+//tag::notable-breaking-changes[]
+==== Java 11 is required
+
+Java 11 is now required to run Elasticsearch and any of its command line tools.
```
Comment (jpountz) [inline docs/reference/migration/migrate_8_0/packaging.asciidoc]: do we need to say "11 or greater" to clarify that more recent versions are fine?

File: server/build.gradle [hunk 2]
```diff
@@ -34,39 +34,39 @@ publishing {
 
 archivesBaseName = 'elasticsearch'
 
-// we want to keep the JDKs in our IDEs set to JDK 8 until minimum JDK is bumped to 9 so we do not include this source set in our IDEs
+// we want to keep the JDKs in our IDEs set to JDK 11 until minimum JDK is bumped to 17 so we do not include this source set in our IDEs
```
Comment (jpountz) [inline server/build.gradle]: s/17/12/ ?

File: server/src/main/java/org/elasticsearch/monitor/jvm/JvmInfo.java [hunk 3]
```diff
@@ -158,10 +158,29 @@
         final boolean bundledJdk = Booleans.parseBoolean(System.getProperty("es.bundled_jdk", Boolean.FALSE.toString()));
         final Boolean usingBundledJdk = bundledJdk ? usingBundledJdk() : null;
 
-        INSTANCE = new JvmInfo(JvmPid.getPid(), System.getProperty("java.version"), runtimeMXBean.getVmName(), runtimeMXBean.getVmVersion(),
-                runtimeMXBean.getVmVendor(), bundledJdk, usingBundledJdk, runtimeMXBean.getStartTime(), configuredInitialHeapSize,
-                configuredMaxHeapSize, mem, inputArguments, bootClassPath, classPath, systemProperties, gcCollectors, memoryPools, onError,
-                onOutOfMemoryError, useCompressedOops, useG1GC, useSerialGC);
+        INSTANCE = new JvmInfo(
+                ProcessHandle.current().pid(),
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/monitor/jvm/JvmInfo.java]: note to other reviewers: this line is the thing that changed

**Review by jpountz — APPROVED** [review]
What I understand looks good, but please wait for @rjernst's review before pushing as I'm not very familiar with the build.

It looks like we could do minor simplifications in `BuildPlugin.groovy` and `PrecommitTasks.groovy`, which check for `project.runtimeJavaVersion >= JavaVersion.VERSION_1_9`, which would always be true now, and `project.compilerJavaVersion < JavaVersion.VERSION_1_10` which would always evaluate to false?

File: .ci/matrix-runtime-javas.yml [hunk 4]
```diff
@@ -6,11 +6,8 @@
 # or 'openjdk' followed by the major release number.
 
 ES_RUNTIME_JAVA:
-  - java8
-  - java8fips
```
Comment (atorok) [inline .ci/matrix-runtime-javas.yml]: Shouldn't we have an alternative before removing this?
There's #37250 so this will not fall trough the cracks, but in the interim we will have no fips testing,
and fips might have to play catch-up once we get the tests back. 
I think that's something we should generally try to avoid in CI.

**Comment by danielmitterdorfer** [comment]
> @danielmitterdorfer This will disrupt performance benchmarks, how do you want to proceed?

Thanks for the ping; highly appreciated. We're all set from a benchmarking perspective and the [cutover benchmarking run](https://elasticsearch-ci.elastic.co/view/All/job/elastic+elasticsearch+master+macrobenchmark-periodic/207/) has just been triggered. From now on all our nightly benchmarks will run with OpenJDK 11.0.2.

File: Vagrantfile [hunk 5]
```diff
@@ -246,6 +246,10 @@ def linux_common(config,
     touch /is_vagrant_vm # for consistency between linux and windows
   SHELL
 
+  config.vm.provision 'jdk-11', type: 'shell', inline: <<-SHELL
+    curl -sSL https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_linux-x64_bin.tar.gz | tar xz -C /opt/
```
Comment (atorok) [inline Vagrantfile]: This can potentially lead to more CI failures due to network issues. 
On the other hand I'm not sure we need to pull in any JDK for the packaging tests,
AFAIK we test with the java version from the platform under test, if this is  to old we should just stick to the bundled one or have the set set JAVA_HOME to that location of that's what we want to test.

File: server/build.gradle (hunk 2 above)
Comment (atorok) [inline server/build.gradle]: Why aren't we removing the additional source set just like we did for libs ? 
Are we anticipating a need to to MR jars for java 12 ?

**Review by alpar-t — COMMENTED** [review]
Thanks! I left some comments and questions.
Note to self, there's some cleanup in `buildSrc/build.gradle` that is made possible by this PR.

File: Vagrantfile [hunk 6]
```diff
@@ -342,10 +346,11 @@ def sh_install_deps(config,
     }
     cat \<\<JAVA > /etc/profile.d/java_home.sh
 if [ -z "\\\$JAVA_HOME" ]; then
-  export JAVA_HOME=$(dirname $(dirname $(readlink -f $(which java))))
+  export JAVA_HOME=/opt/jdk-11.0.2
 fi
 export SYSTEM_JAVA_HOME=\\\$JAVA_HOME
 unset JAVA_HOME
+export PATH=/opt/jdk-11.0.2/bin:\\\$PATH
```
Comment (rjernst) [inline Vagrantfile]: Why is this added to path? The packaging tests don't use java on path anymore.

File: server/build.gradle (hunk 2 above)
Comment (rjernst) [reply server/build.gradle]: I think it should be 18? The next LTS release?

**Review by rjernst — APPROVED** [review]
LGTM

File: server/build.gradle (hunk 2 above)
Comment (jasontedor) [reply server/build.gradle]: Yeah, I was going for the next LTS release which I had in my head was 17. I think that's right though? JDK LTS releases every three years, 11 was in September 2018 so the next one would be September 2021 which would be:
 - 13 September 2019
 - 14 March 2020
 - 15 September 2020
 - 16 March 2021
 - 17 September 2021

Comment (jasontedor) [reply server/build.gradle]: Yeah, I'm anticipating that indeed.

File: Vagrantfile (hunk 5 above)
Comment (jasontedor) [reply Vagrantfile]: We need a non-bundled JDK in the image that can run Elasticsearch to test that a non-bundled JDK works. The installed JDK 8s would be too old, so we need JDK 11.

File: .ci/matrix-runtime-javas.yml (hunk 4 above)
Comment (jasontedor) [reply .ci/matrix-runtime-javas.yml]: I discussed this with @jkakavas and he will be working on getting a FIPS-compliant JDK 11 ready, but agreed that can happen independently. He will focus on that in the upcoming week.

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/2

File: Vagrantfile (hunk 6 above)
Comment (jasontedor) [reply Vagrantfile]: We do use Java on the path. We have lines like:

```
local java=$(which java)
```

I'll rework these tests so that they aren't using Java from the path.

Comment (jasontedor) [reply Vagrantfile]: Here is one example: https://github.com/elastic/elasticsearch/blob/fef21536fa5c0a9ec10e2d715c6bf9452da35258/qa/vagrant/src/test/resources/packaging/tests/module_and_plugin_test_cases.bash#L433
